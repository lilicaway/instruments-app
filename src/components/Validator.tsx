import * as React from 'react';
import { Label } from 'react-bootstrap';

export interface StateProps {
  min: validatorInput;
  max: validatorInput;
  errorMessage: string;
}

type validatorInput = number | Date | '' | undefined;

/** Returns true if all is ok, or false otherwise. */
function validateOverlap(min: validatorInput, max: validatorInput): boolean {
  if (min === '' || min === undefined || max === '' || max === undefined) {
    return true;
  }
  let minValue: number;
  let maxValue: number;
  if (min instanceof Date) {
    minValue = min.getTime();
  } else {
    minValue = min;
  }
  if (max instanceof Date) {
    maxValue = max.getTime();
  } else {
    maxValue = max;
  }
  return maxValue >= minValue;
}

export class Validator extends React.Component<StateProps> {
  public render() {
    if (validateOverlap(this.props.min, this.props.max)) {
      return <span>{''}</span>;
    } else {
      return (
        <Label id="error" bsStyle="danger">
          {this.props.errorMessage}
        </Label>
      );
    }
  }
}
