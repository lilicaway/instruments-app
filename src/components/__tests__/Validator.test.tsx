import * as Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import moment from 'moment';
import * as React from 'react';
import { StateProps, Validator } from '../Validator';

Enzyme.configure({ adapter: new Adapter() });

let component: Enzyme.ShallowWrapper<StateProps, Readonly<{}>, Validator>;
let onSearchChangeFunctionMock: jest.Mock<{}>;

test('Validator for min > max as dates shows error message', () => {
  component = Enzyme.shallow<Validator, StateProps>(
    <Validator
      min={moment('2018-01-01').toDate()}
      max={moment('2014-01-01').toDate()}
      errorMessage="The validator failed with dates"
    />
  );
  expect(component).toMatchSnapshot();
  const elem = component.prop('children');
  expect(elem).toEqual('The validator failed with dates');
});

test('Valdator for min > max as numbers shows error message', () => {
  component = Enzyme.shallow<Validator, StateProps>(
    <Validator
      min={10}
      max={5}
      errorMessage="The validator failed with numbers"
    />
  );
  expect(component).toMatchSnapshot();
  const elem = component.prop('children');
  expect(elem).toEqual('The validator failed with numbers');
});
