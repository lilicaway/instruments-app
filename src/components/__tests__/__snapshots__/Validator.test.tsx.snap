// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Valdator for min > max as numbers shows error message 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Validator
    errorMessage="The validator failed with numbers"
    max={5}
    min={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bsClass": "label",
      "bsStyle": "danger",
      "children": "The validator failed with numbers",
      "id": "error",
    },
    "ref": null,
    "rendered": "The validator failed with numbers",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "label",
        "bsStyle": "danger",
        "children": "The validator failed with numbers",
        "id": "error",
      },
      "ref": null,
      "rendered": "The validator failed with numbers",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Validator for min > max as dates shows error message 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Validator
    errorMessage="The validator failed with dates"
    max={2013-12-31T23:00:00.000Z}
    min={2017-12-31T23:00:00.000Z}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bsClass": "label",
      "bsStyle": "danger",
      "children": "The validator failed with dates",
      "id": "error",
    },
    "ref": null,
    "rendered": "The validator failed with dates",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "label",
        "bsStyle": "danger",
        "children": "The validator failed with dates",
        "id": "error",
      },
      "ref": null,
      "rendered": "The validator failed with dates",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
